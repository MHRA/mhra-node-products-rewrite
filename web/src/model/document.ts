export interface IDocument {
  activeSubstances: string[];
  context: string;
  created: string;
  docType: string;
  fileSize: string;
  name: string;
  product: string;
  url: string;
}

export interface IDocuments {
  count: number;
  documents: IDocument[];
}

export interface IBmgfReport {
  activeSubstances: string[];
  context: string;
  fileName: string;
  fileSize: string;
  fileUrl: string;
  title: string;
  products: string[];
  matrices: string[];
  pbpkModels: string[];
  summary: string;
  url: string;
  pregnancyTrimesters: string[];
  plNumbers: string[];
}

export interface IBmgfReports {
  count: number;
  reports: IBmgfReport[];
}
