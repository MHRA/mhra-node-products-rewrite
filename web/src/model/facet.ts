export interface IFacet {
  name: string;
  count?: number;
}
