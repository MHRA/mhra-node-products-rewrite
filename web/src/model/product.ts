export interface IProduct {
  name: string;
  count?: number;
}
