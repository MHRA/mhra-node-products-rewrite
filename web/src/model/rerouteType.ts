export enum RerouteType {
  CheckboxSelected,
  Other,
}
