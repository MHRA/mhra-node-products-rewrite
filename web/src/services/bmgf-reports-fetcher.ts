import {
  BlobServiceClient,
  StorageSharedKeyCredential,
} from '@azure/storage-blob';

const containerUrl = `https://${process.env.AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/bmgf-docs`;
const account = process.env.AZURE_STORAGE_ACCOUNT || '';
const accountKey = process.env.AZURE_STORAGE_KEY || '';

export const getReportList = async (): Promise<string[]> => {
  const sharedKeyCredential = new StorageSharedKeyCredential(
    account,
    accountKey,
  );
  const blobServiceClient = new BlobServiceClient(
    `https://${account}.blob.core.windows.net`,
    sharedKeyCredential,
  );

  const containerName = 'bmgf-docs';
  const containerClient = blobServiceClient.getContainerClient(containerName);

  const blobs = containerClient.listBlobsFlat();
  const reportNames = [];
  for await (const blob of blobs) {
    if (blob.name.endsWith('.html')) {
      reportNames.push(blob.name);
    }
  }
  return reportNames;
};

export const getReportUrls = async (
  reportName: string,
  fileNames: string[],
): Promise<any> => {
  const fileName = fileNames.find((fileName) =>
    fileName.startsWith(reportName),
  );
  return {
    reportPdfUrl: `${containerUrl}/${fileName?.replace('.html', '.pdf')}`,
    reportHtmlUrl: `${containerUrl}/${fileName}`,
    reportAssetsUrl: `${containerUrl}/${reportName}/assets/`,
  };
};

export const getReportHtmlContent = async (
  reportUrl: string,
): Promise<string> => {
  return fetch(`${reportUrl}`, {
    method: 'GET',
    headers: {
      'Content-Type': 'text/html',
    },
  }).then((response) => (response.ok ? response.text() : ''));
};
