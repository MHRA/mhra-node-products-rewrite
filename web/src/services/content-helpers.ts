export const pluralise = (
  singular: string,
  plural: string,
  count: number,
): string => {
  return count === 1 ? singular : plural;
};
