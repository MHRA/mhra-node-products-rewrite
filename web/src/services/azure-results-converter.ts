import moment from 'moment';

import { IDocument, IBmgfReport } from '../model/document';
import { IFacet } from '../model/facet';
import {
  IFacetResult,
  ISearchResult,
  IBmgfSearchResult,
} from '../services/azure-search';

const sanitizeTitle = (title: string | null): string => {
  let name: string;
  if (!title) return 'Unknown';

  try {
    name = decodeURIComponent(title);
  } catch {
    name = title;
  }
  return name;
};

export const convertResults = (doc: ISearchResult): IDocument => {
  return {
    activeSubstances: doc.substance_name,
    product: doc.product_name,
    context: doc['@search.highlights']?.content.join(' … ') || '',
    docType: doc.doc_type?.toString().substr(0, 3) || '',
    fileSize: Math.ceil(
      (doc.metadata_storage_size ? doc.metadata_storage_size : 0) / 1000,
    ).toLocaleString('en-GB'),
    created: doc.created
      ? moment(doc.created).format('D MMMM YYYY')
      : 'Unknown',
    name: sanitizeTitle(doc.title),
    url: doc.metadata_storage_path,
  };
};

export const convertBmgfResults = (doc: IBmgfSearchResult): IBmgfReport => {
  return {
    activeSubstances: doc.active_substances,
    context: doc['@search.highlights']?.content.join(' … ') || '',
    fileName: doc.file_name || '',
    fileUrl: doc.metadata_storage_path,
    products: doc.products,
    summary: doc.summary,
    pbpkModels: doc.pbpk_models,
    matrices: doc.matrices,
    title: sanitizeTitle(doc.report_name),
    fileSize: Math.ceil(
      (doc.metadata_storage_size ? doc.metadata_storage_size : 0) / 1000,
    ).toLocaleString('en-GB'),
    url: `/medicine-levels-in-pregnancy/reports/${doc.report_name}`,
    plNumbers: doc.pl_numbers,
    pregnancyTrimesters: doc.pregnancy_trimesters,
  };
};

export const mapSubstancesIndex = ([letter, facetResult]: [
  string,
  IFacetResult,
]): IFacet[] => {
  const indexResults: { [id: string]: IFacet } = {};
  facetResult.facets
    .filter((x) => x.value.startsWith(letter))
    .forEach((f) => {
      const [substance, product] = f.value
        .replace(/\s+/g, ' ')
        .split(', ', 3)
        .slice(1);

      if (substance && !product) {
        const currentSubstanceCount = indexResults[substance]?.count ?? 0;
        indexResults[substance] = {
          name: substance,
          count: currentSubstanceCount + f.count,
        };
      }
    });
  return Object.values(indexResults).sort((a, b) => (a.name < b.name ? -1 : 1));
};

export const mapProductsIndex = ([letterAndSubstance, facetResult]: [
  string,
  IFacetResult,
]): IFacet[] => {
  const indexResults: { [id: string]: IFacet } = {};
  const requiredFacetPrefix = `${letterAndSubstance},`;
  facetResult.facets.forEach((f) => {
    const sanitizedFacet = f.value.replace(/\s+/g, ' ');
    if (!sanitizedFacet.startsWith(requiredFacetPrefix)) {
      return;
    }

    const [product] = sanitizedFacet.split(', ', 3).slice(2);

    if (product) {
      const currentProductCount = indexResults[product]?.count ?? 0;
      indexResults[product] = {
        name: product,
        count: currentProductCount + f.count,
      };
    }
  });
  return Object.values(indexResults).sort((a, b) => (a.name < b.name ? -1 : 1));
};
