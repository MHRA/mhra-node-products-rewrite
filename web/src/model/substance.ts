import { IProduct } from './product';

export interface ISubstance extends IProduct {
  products?: IProduct[];
}

export function isSubstance(obj: any): obj is ISubstance {
  return obj.products !== undefined && obj.count !== undefined;
}

export function isIndex(obj: any): obj is ISubstance {
  return obj.products === undefined && obj.count === undefined;
}
