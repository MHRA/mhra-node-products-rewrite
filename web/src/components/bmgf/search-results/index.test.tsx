import renderer from 'react-test-renderer';
import React from 'react';
import { RerouteType } from '../../../model/rerouteType';
import { DocType } from '../../../services/azure-search';
import SearchResults from './index';

const drugsMock = [
  {
    activeSubstances: ['substance 1', 'substance 2'],
    context: 'context',
    fileName: 'file name',
    fileSize: 'file size',
    fileUrl: 'file url',
    title: 'title',
    products: ['product 1', 'product 2'],
    matrices: ['matrix 1', 'matrix 2'],
    pbpkModels: ['model 1', 'model 2'],
    summary: 'summary',
    url: 'url',
    pregnancyTrimesters: ['first', 'second'],
    plNumbers: ['PL123451234', 'PL234562345'],
  },
  {
    activeSubstances: ['substance 3', 'substance 4'],
    context: 'context',
    fileName: 'file name',
    fileSize: 'file size',
    fileUrl: 'file url',
    title: 'title 2',
    products: ['product 3', 'product 4'],
    matrices: ['matrix 3', 'matrix 4'],
    pbpkModels: ['model 3', 'model 4'],
    summary: 'summary',
    url: 'url',
    pregnancyTrimesters: ['third'],
    plNumbers: [],
  },
];

const noFeedback = () => undefined;
const updateDocType = (d: DocType[]) => undefined;

describe(SearchResults, () => {
  it('should render', () => {
    const component = renderer
      .create(
        <SearchResults
          reports={drugsMock}
          page={1}
          pageSize={20}
          resultCount={200}
          searchTerm={'Tea'}
          showingResultsForTerm={'Tea'}
          handlePageChange={noFeedback}
          isLoading={false}
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should render loading page', () => {
    const component = renderer
      .create(
        <SearchResults
          reports={drugsMock}
          page={1}
          pageSize={20}
          resultCount={200}
          searchTerm={'Tea'}
          showingResultsForTerm={'Tea'}
          handlePageChange={noFeedback}
          isLoading
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
