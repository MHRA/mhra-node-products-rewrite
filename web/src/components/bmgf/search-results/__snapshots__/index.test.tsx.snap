// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults should render 1`] = `
Array [
  <div
    className="sc-gsnTZi hmKVXy"
  >
    <div>
      <div
        className="sc-eCYdqJ gZAhPP"
      >
        <h2
          className="title"
        >
          Showing results for Tea
        </h2>
        <p
          className="sc-jSMfEi iCawsZ no-of-results"
        >
          1 to 2 of 200
        </p>
      </div>
      <p>
        If you’re looking for other information on medicines, you can search for information about medicines including patient information leaflets (PILs), details on how the medicine can be used (SPCs) and scientific reports (PARs).
      </p>
      <p>
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Go to Products website to find information on medicines
        </a>
      </p>
    </div>
    <div
      className="row"
    >
      <section
        className="column results"
      >
        <h3
          className="sc-dkzDqf eoxGfn"
        >
          Search results
        </h3>
        <dl>
          <div
            className="search-result"
          >
            <dd
              className="right"
            >
              <a
                href="url"
              >
                <p
                  className="title"
                >
                  title
                </p>
              </a>
              <p
                className="metadata"
              >
                Active substances:
                 
                Substance 1, Substance 2
              </p>
              <p
                className="summary"
              >
                summary
              </p>
            </dd>
          </div>
          <div
            className="search-result"
          >
            <dd
              className="right"
            >
              <a
                href="url"
              >
                <p
                  className="title"
                >
                  title 2
                </p>
              </a>
              <p
                className="metadata"
              >
                Active substances:
                 
                Substance 3, Substance 4
              </p>
              <p
                className="summary"
              >
                summary
              </p>
            </dd>
          </div>
        </dl>
      </section>
    </div>
  </div>,
  <nav
    className="sc-bczRLJ wtnsX"
  >
    <nav
      aria-label="Pagination Navigation"
    >
      <div
        className="pagination"
      >
        <ul
          className="pagination-number"
        >
          <li
            aria-current="true"
            aria-label="Current Page, Page 1"
          >
            1
            <span>
              –
            </span>
          </li>
          <li
            aria-label="Goto Page 2"
            onClick={[Function]}
          >
            <span
              className="link-text"
            >
              2
            </span>
          </li>
          <li>
            …
          </li>
          <li
            aria-label="Goto Page 11"
            onClick={[Function]}
          >
            <span
              className="link-text"
            >
              11
            </span>
          </li>
        </ul>
        <button
          aria-label="Goto next page, Page 2"
          className="arrow"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </nav>
  </nav>,
]
`;

exports[`SearchResults should render loading page 1`] = `
<div
  className="sc-gsnTZi hmKVXy"
>
  <h2
    className="title"
  >
    Loading results for Tea...
  </h2>
</div>
`;
